src/tools/r2rdump/R2RReader.cs

```diff
@@ -113,35 +112,77 @@ public unsafe R2RReader(string filename)
+                            for (int i = 0; i < nArgs; i++)
+                            {
+                                args[i] = (R2RMethod.GenericElementTypes)curEntry.GetByte();
+                                if (args[i] == R2RMethod.GenericElementTypes.ValueType)
+                                {
+                                    tokens[i] = curEntry.GetByte();
+                                    tokens[i] = (tokens[i] >> 2);
+                                }
+                            }
+
+                            uint id = curEntry.GetUnsigned();
+                            id = id >> 1;
+                            R2RMethod method = new R2RMethod(_image, mdReader, rid, (int)id, args, tokens);
+                            if (method.EntryPointRuntimeFunctionId >= 0 && method.EntryPointRuntimeFunctionId < nRuntimeFunctions)
+                            {
+                                isEntryPoint[method.EntryPointRuntimeFunctionId] = true;
+                            }
+                            R2RMethods.Add(method);
+                        }
+                        curEntry = allEntries.GetNext();
+                    }
+
                     // get the RVAs of the runtime functions for each method
-                    int runtimeFunctionId = 0;
+                    int curOffset = 0;
                     foreach (R2RMethod method in R2RMethods)
                     {
-                        int curOffset = (int)(runtimeFunctionOffset + method.EntryPointRuntimeFunctionId * runtimeFunctionSize);
+                        int runtimeFunctionId = method.EntryPointRuntimeFunctionId;
+                        if (runtimeFunctionId == -1)
+                            continue;
+                        curOffset = runtimeFunctionOffset + runtimeFunctionId * runtimeFunctionSize;
                         do
                         {
-                            int startRva = ReadInt32(_image, ref curOffset);
+                            int startRva = NativeReader.ReadInt32(_image, ref curOffset);
                             int endRva = -1;
                             if (Machine == Machine.Amd64)
                             {
-                                endRva = ReadInt32(_image, ref curOffset);
+                                endRva = NativeReader.ReadInt32(_image, ref curOffset);
                             }
-                            int unwindRva = ReadInt32(_image, ref curOffset);
+                            int unwindRva = NativeReader.ReadInt32(_image, ref curOffset);
 
-                            method.NativeCode.Add(new RuntimeFunction(startRva, endRva, unwindRva));
+                            method.NativeCode.Add(new RuntimeFunction(runtimeFunctionId, startRva, endRva, unwindRva));
                             runtimeFunctionId++;
                         }
                         while (runtimeFunctionId < nRuntimeFunctions && !isEntryPoint[runtimeFunctionId]);
```

Comment:
Was there a reason this was a do while loop instead of a regular while loop?

Reply:
I used do while because the first runtimeFunctionId is an entrypoint. Would it be better to use a regular while loop and check if it's entrypoint at the end and break?

Reply:
Oh I see that now! No, this do while should be fine. Thanks for pointing that out.